{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Cut Out UTRs\n",
    "Code to cut out the UTRs out of the sequences of the intergenic sequences"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Imports"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import numpy as np\n",
    "import matplotlib.pyplot as plt"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<a id=\"help\"></a>\n",
    "### General Helper Functions"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "def rev_comp(seq):\n",
    "    '''\n",
    "    Switches a seqence from top to bottom \n",
    "    (ie 'ATGC' becomes 'TACG')\n",
    "    \n",
    "    Args:\n",
    "        seq: a string which represents the top strand of DNA.\n",
    "        Should only have the characters ATGC\n",
    "    Ret:\n",
    "        string sequence that represents the bottom sequence\n",
    "        (all Ts become As, As become Ts, Gs become Cs, Cs become Gs)\n",
    "        and is revered\n",
    "        Resulting string will be all uppercase\n",
    "    '''\n",
    "    res = seq.lower()\n",
    "    res = res.replace('t', 'A')\n",
    "    res = res.replace('a', 'T')\n",
    "    res = res.replace('g', 'C')\n",
    "    res = res.replace('c', 'G')\n",
    "    res = res[::-1]\n",
    "    return res\n",
    "\n",
    "def comp(seq):\n",
    "    '''\n",
    "    Returns the compliment of this sequence\n",
    "    (all Ts become As, As become Ts, Gs become Cs, Cs become Gs)\n",
    "    Us all uppercase\n",
    "    '''\n",
    "    res = seq.lower()\n",
    "    res = res.replace('t', 'A')\n",
    "    res = res.replace('a', 'T')\n",
    "    res = res.replace('g', 'C')\n",
    "    res = res.replace('c', 'G')\n",
    "    return res\n",
    "\n",
    "def rev(seq):\n",
    "    '''\n",
    "    reverses a string\n",
    "    '''\n",
    "    res = seq[::-1]\n",
    "    return res"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<a id=\"load\"></a>\n",
    "### Loading in Data\n",
    "We load in our two data sets.\n",
    "\n",
    "- df_intergenic contains the data we cleaned regarding the intergenic sequences\n",
    "- df_TSS contains TSS positions and UTR lengths"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "df_intergenic = pd.read_excel('./source/Upstream-Sequences_df.xls')\n",
    "df_TSS = pd.read_excel('./source/journal.pgen.1002867.s007.XLSX')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Unnamed: 0</th>\n",
       "      <th>IS Number</th>\n",
       "      <th>Gene</th>\n",
       "      <th>Upstream Sequence</th>\n",
       "      <th>L_END</th>\n",
       "      <th>R_END</th>\n",
       "      <th>Length</th>\n",
       "      <th>Adjacent Gene</th>\n",
       "      <th>Orientation</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>3</td>\n",
       "      <td>yaaX</td>\n",
       "      <td>ATCTATTCATTATCTCAATCAGGCCGGGTTTGCTTTTATGCAGCCC...</td>\n",
       "      <td>5021</td>\n",
       "      <td>5233</td>\n",
       "      <td>213</td>\n",
       "      <td>thrC</td>\n",
       "      <td>Codirectional+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>5</td>\n",
       "      <td>yaaA</td>\n",
       "      <td>ATCACAACTATCGATCAACTCATTCTCATTTTTTGCTAAAGTCGGC...</td>\n",
       "      <td>6460</td>\n",
       "      <td>6528</td>\n",
       "      <td>69</td>\n",
       "      <td>yaaJ</td>\n",
       "      <td>Codirectional-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>6</td>\n",
       "      <td>yaaJ</td>\n",
       "      <td>TGATAGTATTTCTCTTTAAACAGCTTGTTAGGGGGATGTAACCGGT...</td>\n",
       "      <td>7960</td>\n",
       "      <td>8237</td>\n",
       "      <td>278</td>\n",
       "      <td>talB</td>\n",
       "      <td>Divergent</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>3</td>\n",
       "      <td>3</td>\n",
       "      <td>6</td>\n",
       "      <td>talB</td>\n",
       "      <td>CATATCCCTCTTATTGCCGGTCGCGATGACTTTCCTGTGTAAACGT...</td>\n",
       "      <td>7960</td>\n",
       "      <td>8237</td>\n",
       "      <td>278</td>\n",
       "      <td>yaaJ</td>\n",
       "      <td>Divergent</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>4</td>\n",
       "      <td>4</td>\n",
       "      <td>7</td>\n",
       "      <td>mog</td>\n",
       "      <td>CATTCTTAGCGTGACCGGGAAGTCGGTCACGCTACCTCTTCTGAAG...</td>\n",
       "      <td>9192</td>\n",
       "      <td>9305</td>\n",
       "      <td>114</td>\n",
       "      <td>talB</td>\n",
       "      <td>Codirectional+</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   Unnamed: 0  IS Number  Gene  \\\n",
       "0           0          3  yaaX   \n",
       "1           1          5  yaaA   \n",
       "2           2          6  yaaJ   \n",
       "3           3          6  talB   \n",
       "4           4          7   mog   \n",
       "\n",
       "                                   Upstream Sequence  L_END  R_END  Length  \\\n",
       "0  ATCTATTCATTATCTCAATCAGGCCGGGTTTGCTTTTATGCAGCCC...   5021   5233     213   \n",
       "1  ATCACAACTATCGATCAACTCATTCTCATTTTTTGCTAAAGTCGGC...   6460   6528      69   \n",
       "2  TGATAGTATTTCTCTTTAAACAGCTTGTTAGGGGGATGTAACCGGT...   7960   8237     278   \n",
       "3  CATATCCCTCTTATTGCCGGTCGCGATGACTTTCCTGTGTAAACGT...   7960   8237     278   \n",
       "4  CATTCTTAGCGTGACCGGGAAGTCGGTCACGCTACCTCTTCTGAAG...   9192   9305     114   \n",
       "\n",
       "  Adjacent Gene     Orientation  \n",
       "0          thrC  Codirectional+  \n",
       "1          yaaJ  Codirectional-  \n",
       "2          talB       Divergent  \n",
       "3          yaaJ       Divergent  \n",
       "4          talB  Codirectional+  "
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#df_intergenic contains data that we cleaned regarding intergenic sequences\n",
    "df_intergenic.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>TSS position</th>\n",
       "      <th>TSS reads</th>\n",
       "      <th>Strand</th>\n",
       "      <th>Gene ID</th>\n",
       "      <th>Gene name</th>\n",
       "      <th>Gene product</th>\n",
       "      <th>COG function</th>\n",
       "      <th>Gene start</th>\n",
       "      <th>Gene end</th>\n",
       "      <th>5' UTR length</th>\n",
       "      <th>Sequence -50 nt upstream + TSS (51 nt)</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <td>0</td>\n",
       "      <td>NC_000913</td>\n",
       "      <td>148</td>\n",
       "      <td>1475</td>\n",
       "      <td>+</td>\n",
       "      <td>b0001</td>\n",
       "      <td>thrL</td>\n",
       "      <td>thr operon leader peptide</td>\n",
       "      <td>Not in COGs</td>\n",
       "      <td>190</td>\n",
       "      <td>255</td>\n",
       "      <td>42</td>\n",
       "      <td>AATTAAAATTTTATTGACTTAGGTCACTAAATACTTTAACCAATAT...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>1</td>\n",
       "      <td>NC_000913</td>\n",
       "      <td>195</td>\n",
       "      <td>11</td>\n",
       "      <td>+</td>\n",
       "      <td>b0002</td>\n",
       "      <td>thrA</td>\n",
       "      <td>aspartate kinase / homoserine dehydrogenase</td>\n",
       "      <td>Amino acid transport and metabolism</td>\n",
       "      <td>337</td>\n",
       "      <td>2799</td>\n",
       "      <td>142</td>\n",
       "      <td>GGCATAGCGCACAGACAGATAAAAATTACAGAGTACACAACATCCA...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>2</td>\n",
       "      <td>NC_000913</td>\n",
       "      <td>2391</td>\n",
       "      <td>3</td>\n",
       "      <td>+</td>\n",
       "      <td>b0003</td>\n",
       "      <td>thrB</td>\n",
       "      <td>homoserine kinase</td>\n",
       "      <td>Amino acid transport and metabolism</td>\n",
       "      <td>2801</td>\n",
       "      <td>3733</td>\n",
       "      <td>410</td>\n",
       "      <td>GACCCGCGAGATGATCTTTCTGGTATGGATGTGGCGCGTAAACTAT...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>3</td>\n",
       "      <td>NC_000913</td>\n",
       "      <td>2693</td>\n",
       "      <td>6</td>\n",
       "      <td>+</td>\n",
       "      <td>b0003</td>\n",
       "      <td>thrB</td>\n",
       "      <td>homoserine kinase</td>\n",
       "      <td>Amino acid transport and metabolism</td>\n",
       "      <td>2801</td>\n",
       "      <td>3733</td>\n",
       "      <td>108</td>\n",
       "      <td>CAAAGTGAAAAATGGCGAAAACGCCCTGGCCTTCTATAGCCACTAT...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>4</td>\n",
       "      <td>NC_000913</td>\n",
       "      <td>2742</td>\n",
       "      <td>4</td>\n",
       "      <td>+</td>\n",
       "      <td>b0003</td>\n",
       "      <td>thrB</td>\n",
       "      <td>homoserine kinase</td>\n",
       "      <td>Amino acid transport and metabolism</td>\n",
       "      <td>2801</td>\n",
       "      <td>3733</td>\n",
       "      <td>59</td>\n",
       "      <td>CAGCCGCTGCCGTTGGTACTGCGCGGATATGGTGCGGGCAATGACG...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  Chromosome  TSS position  TSS reads Strand Gene ID Gene name  \\\n",
       "0  NC_000913           148       1475      +   b0001      thrL   \n",
       "1  NC_000913           195         11      +   b0002      thrA   \n",
       "2  NC_000913          2391          3      +   b0003      thrB   \n",
       "3  NC_000913          2693          6      +   b0003      thrB   \n",
       "4  NC_000913          2742          4      +   b0003      thrB   \n",
       "\n",
       "                                  Gene product  \\\n",
       "0                    thr operon leader peptide   \n",
       "1  aspartate kinase / homoserine dehydrogenase   \n",
       "2                            homoserine kinase   \n",
       "3                            homoserine kinase   \n",
       "4                            homoserine kinase   \n",
       "\n",
       "                          COG function  Gene start  Gene end  5' UTR length  \\\n",
       "0                          Not in COGs         190       255             42   \n",
       "1  Amino acid transport and metabolism         337      2799            142   \n",
       "2  Amino acid transport and metabolism        2801      3733            410   \n",
       "3  Amino acid transport and metabolism        2801      3733            108   \n",
       "4  Amino acid transport and metabolism        2801      3733             59   \n",
       "\n",
       "              Sequence -50 nt upstream + TSS (51 nt)  \n",
       "0  AATTAAAATTTTATTGACTTAGGTCACTAAATACTTTAACCAATAT...  \n",
       "1  GGCATAGCGCACAGACAGATAAAAATTACAGAGTACACAACATCCA...  \n",
       "2  GACCCGCGAGATGATCTTTCTGGTATGGATGTGGCGCGTAAACTAT...  \n",
       "3  CAAAGTGAAAAATGGCGAAAACGCCCTGGCCTTCTATAGCCACTAT...  \n",
       "4  CAGCCGCTGCCGTTGGTACTGCGCGGATATGGTGCGGGCAATGACG...  "
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#df_TSS contains data regarding TSS positions and UTR lengths\n",
    "df_TSS.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "---\n",
    "<a id=\"clean\"></a>\n",
    "## Cleaning Data\n",
    "<a id=\"merge\"></a>\n",
    "### Merging datasets"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Unnamed: 0</th>\n",
       "      <th>IS Number</th>\n",
       "      <th>Gene</th>\n",
       "      <th>Upstream Sequence</th>\n",
       "      <th>L_END</th>\n",
       "      <th>R_END</th>\n",
       "      <th>Length</th>\n",
       "      <th>Adjacent Gene</th>\n",
       "      <th>Orientation</th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>...</th>\n",
       "      <th>TSS reads</th>\n",
       "      <th>Strand</th>\n",
       "      <th>Gene ID</th>\n",
       "      <th>Gene name</th>\n",
       "      <th>Gene product</th>\n",
       "      <th>COG function</th>\n",
       "      <th>Gene start</th>\n",
       "      <th>Gene end</th>\n",
       "      <th>5' UTR length</th>\n",
       "      <th>Sequence -50 nt upstream + TSS (51 nt)</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>3</td>\n",
       "      <td>yaaX</td>\n",
       "      <td>ATCTATTCATTATCTCAATCAGGCCGGGTTTGCTTTTATGCAGCCC...</td>\n",
       "      <td>5021</td>\n",
       "      <td>5233</td>\n",
       "      <td>213</td>\n",
       "      <td>thrC</td>\n",
       "      <td>Codirectional+</td>\n",
       "      <td>NC_000913</td>\n",
       "      <td>...</td>\n",
       "      <td>2</td>\n",
       "      <td>+</td>\n",
       "      <td>b0005</td>\n",
       "      <td>yaaX</td>\n",
       "      <td>predicted protein</td>\n",
       "      <td>Not in COGs</td>\n",
       "      <td>5234</td>\n",
       "      <td>5530</td>\n",
       "      <td>448</td>\n",
       "      <td>ACCACGCAACAGACAATGCGTGAGTTAAAAGAACTGGGCTACACTT...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>3</td>\n",
       "      <td>yaaX</td>\n",
       "      <td>ATCTATTCATTATCTCAATCAGGCCGGGTTTGCTTTTATGCAGCCC...</td>\n",
       "      <td>5021</td>\n",
       "      <td>5233</td>\n",
       "      <td>213</td>\n",
       "      <td>thrC</td>\n",
       "      <td>Codirectional+</td>\n",
       "      <td>NC_000913</td>\n",
       "      <td>...</td>\n",
       "      <td>2</td>\n",
       "      <td>+</td>\n",
       "      <td>b0005</td>\n",
       "      <td>yaaX</td>\n",
       "      <td>predicted protein</td>\n",
       "      <td>Not in COGs</td>\n",
       "      <td>5234</td>\n",
       "      <td>5530</td>\n",
       "      <td>354</td>\n",
       "      <td>ATCCAGGCGAATATGGCTTGTTCCTCGGCACCGCGCATCCGGCGAA...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>3</td>\n",
       "      <td>yaaX</td>\n",
       "      <td>ATCTATTCATTATCTCAATCAGGCCGGGTTTGCTTTTATGCAGCCC...</td>\n",
       "      <td>5021</td>\n",
       "      <td>5233</td>\n",
       "      <td>213</td>\n",
       "      <td>thrC</td>\n",
       "      <td>Codirectional+</td>\n",
       "      <td>NC_000913</td>\n",
       "      <td>...</td>\n",
       "      <td>4</td>\n",
       "      <td>+</td>\n",
       "      <td>b0005</td>\n",
       "      <td>yaaX</td>\n",
       "      <td>predicted protein</td>\n",
       "      <td>Not in COGs</td>\n",
       "      <td>5234</td>\n",
       "      <td>5530</td>\n",
       "      <td>346</td>\n",
       "      <td>GAATATGGCTTGTTCCTCGGCACCGCGCATCCGGCGAAATTTAAAG...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>3</td>\n",
       "      <td>0</td>\n",
       "      <td>3</td>\n",
       "      <td>yaaX</td>\n",
       "      <td>ATCTATTCATTATCTCAATCAGGCCGGGTTTGCTTTTATGCAGCCC...</td>\n",
       "      <td>5021</td>\n",
       "      <td>5233</td>\n",
       "      <td>213</td>\n",
       "      <td>thrC</td>\n",
       "      <td>Codirectional+</td>\n",
       "      <td>NC_000913</td>\n",
       "      <td>...</td>\n",
       "      <td>4</td>\n",
       "      <td>+</td>\n",
       "      <td>b0005</td>\n",
       "      <td>yaaX</td>\n",
       "      <td>predicted protein</td>\n",
       "      <td>Not in COGs</td>\n",
       "      <td>5234</td>\n",
       "      <td>5530</td>\n",
       "      <td>303</td>\n",
       "      <td>AAGAGAGCGTGGAAGCGATTCTCGGTGAAACGTTGGATCTGCCAAA...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>4</td>\n",
       "      <td>2</td>\n",
       "      <td>6</td>\n",
       "      <td>yaaJ</td>\n",
       "      <td>TGATAGTATTTCTCTTTAAACAGCTTGTTAGGGGGATGTAACCGGT...</td>\n",
       "      <td>7960</td>\n",
       "      <td>8237</td>\n",
       "      <td>278</td>\n",
       "      <td>talB</td>\n",
       "      <td>Divergent</td>\n",
       "      <td>NC_000913</td>\n",
       "      <td>...</td>\n",
       "      <td>15</td>\n",
       "      <td>-</td>\n",
       "      <td>b0007</td>\n",
       "      <td>yaaJ</td>\n",
       "      <td>YaaJ alanine AGSS transporter</td>\n",
       "      <td>Amino acid transport and metabolism</td>\n",
       "      <td>6529</td>\n",
       "      <td>7959</td>\n",
       "      <td>209</td>\n",
       "      <td>CAGCTTGTTAGGGGGATGTAACCGGTCTGCCCTGATGATATCACGA...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 21 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "   Unnamed: 0  IS Number  Gene  \\\n",
       "0           0          3  yaaX   \n",
       "1           0          3  yaaX   \n",
       "2           0          3  yaaX   \n",
       "3           0          3  yaaX   \n",
       "4           2          6  yaaJ   \n",
       "\n",
       "                                   Upstream Sequence  L_END  R_END  Length  \\\n",
       "0  ATCTATTCATTATCTCAATCAGGCCGGGTTTGCTTTTATGCAGCCC...   5021   5233     213   \n",
       "1  ATCTATTCATTATCTCAATCAGGCCGGGTTTGCTTTTATGCAGCCC...   5021   5233     213   \n",
       "2  ATCTATTCATTATCTCAATCAGGCCGGGTTTGCTTTTATGCAGCCC...   5021   5233     213   \n",
       "3  ATCTATTCATTATCTCAATCAGGCCGGGTTTGCTTTTATGCAGCCC...   5021   5233     213   \n",
       "4  TGATAGTATTTCTCTTTAAACAGCTTGTTAGGGGGATGTAACCGGT...   7960   8237     278   \n",
       "\n",
       "  Adjacent Gene     Orientation Chromosome  ...  TSS reads  Strand Gene ID  \\\n",
       "0          thrC  Codirectional+  NC_000913  ...          2       +   b0005   \n",
       "1          thrC  Codirectional+  NC_000913  ...          2       +   b0005   \n",
       "2          thrC  Codirectional+  NC_000913  ...          4       +   b0005   \n",
       "3          thrC  Codirectional+  NC_000913  ...          4       +   b0005   \n",
       "4          talB       Divergent  NC_000913  ...         15       -   b0007   \n",
       "\n",
       "  Gene name                   Gene product  \\\n",
       "0      yaaX              predicted protein   \n",
       "1      yaaX              predicted protein   \n",
       "2      yaaX              predicted protein   \n",
       "3      yaaX              predicted protein   \n",
       "4      yaaJ  YaaJ alanine AGSS transporter   \n",
       "\n",
       "                          COG function Gene start  Gene end  5' UTR length  \\\n",
       "0                          Not in COGs       5234      5530            448   \n",
       "1                          Not in COGs       5234      5530            354   \n",
       "2                          Not in COGs       5234      5530            346   \n",
       "3                          Not in COGs       5234      5530            303   \n",
       "4  Amino acid transport and metabolism       6529      7959            209   \n",
       "\n",
       "              Sequence -50 nt upstream + TSS (51 nt)  \n",
       "0  ACCACGCAACAGACAATGCGTGAGTTAAAAGAACTGGGCTACACTT...  \n",
       "1  ATCCAGGCGAATATGGCTTGTTCCTCGGCACCGCGCATCCGGCGAA...  \n",
       "2  GAATATGGCTTGTTCCTCGGCACCGCGCATCCGGCGAAATTTAAAG...  \n",
       "3  AAGAGAGCGTGGAAGCGATTCTCGGTGAAACGTTGGATCTGCCAAA...  \n",
       "4  CAGCTTGTTAGGGGGATGTAACCGGTCTGCCCTGATGATATCACGA...  \n",
       "\n",
       "[5 rows x 21 columns]"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# merged dataframe, merging on the Gene names\n",
    "df_merged = df_intergenic.merge(df_TSS, left_on='Gene', right_on='Gene name')\n",
    "# df_merged_cut = df_merged_cut.drop(\"Gene name\", axis=1)\n",
    "df_merged.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(2058, 2520, 3746)"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(df_merged), len(df_intergenic), len(df_TSS)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<a id=\"cut\"></a>\n",
    "### Cutting UTRs\n",
    "We cut the UTRs for sequence"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Unnamed: 0</th>\n",
       "      <th>IS Number</th>\n",
       "      <th>Gene</th>\n",
       "      <th>Upstream Sequence</th>\n",
       "      <th>L_END</th>\n",
       "      <th>R_END</th>\n",
       "      <th>Length</th>\n",
       "      <th>Adjacent Gene</th>\n",
       "      <th>Orientation</th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>...</th>\n",
       "      <th>Strand</th>\n",
       "      <th>Gene ID</th>\n",
       "      <th>Gene name</th>\n",
       "      <th>Gene product</th>\n",
       "      <th>COG function</th>\n",
       "      <th>Gene start</th>\n",
       "      <th>Gene end</th>\n",
       "      <th>5' UTR length</th>\n",
       "      <th>Sequence -50 nt upstream + TSS (51 nt)</th>\n",
       "      <th>Sequence without UTR</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>3</td>\n",
       "      <td>yaaX</td>\n",
       "      <td>ATCTATTCATTATCTCAATCAGGCCGGGTTTGCTTTTATGCAGCCC...</td>\n",
       "      <td>5021</td>\n",
       "      <td>5233</td>\n",
       "      <td>213</td>\n",
       "      <td>thrC</td>\n",
       "      <td>Codirectional+</td>\n",
       "      <td>NC_000913</td>\n",
       "      <td>...</td>\n",
       "      <td>+</td>\n",
       "      <td>b0005</td>\n",
       "      <td>yaaX</td>\n",
       "      <td>predicted protein</td>\n",
       "      <td>Not in COGs</td>\n",
       "      <td>5234</td>\n",
       "      <td>5530</td>\n",
       "      <td>448</td>\n",
       "      <td>ACCACGCAACAGACAATGCGTGAGTTAAAAGAACTGGGCTACACTT...</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>3</td>\n",
       "      <td>yaaX</td>\n",
       "      <td>ATCTATTCATTATCTCAATCAGGCCGGGTTTGCTTTTATGCAGCCC...</td>\n",
       "      <td>5021</td>\n",
       "      <td>5233</td>\n",
       "      <td>213</td>\n",
       "      <td>thrC</td>\n",
       "      <td>Codirectional+</td>\n",
       "      <td>NC_000913</td>\n",
       "      <td>...</td>\n",
       "      <td>+</td>\n",
       "      <td>b0005</td>\n",
       "      <td>yaaX</td>\n",
       "      <td>predicted protein</td>\n",
       "      <td>Not in COGs</td>\n",
       "      <td>5234</td>\n",
       "      <td>5530</td>\n",
       "      <td>354</td>\n",
       "      <td>ATCCAGGCGAATATGGCTTGTTCCTCGGCACCGCGCATCCGGCGAA...</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>3</td>\n",
       "      <td>yaaX</td>\n",
       "      <td>ATCTATTCATTATCTCAATCAGGCCGGGTTTGCTTTTATGCAGCCC...</td>\n",
       "      <td>5021</td>\n",
       "      <td>5233</td>\n",
       "      <td>213</td>\n",
       "      <td>thrC</td>\n",
       "      <td>Codirectional+</td>\n",
       "      <td>NC_000913</td>\n",
       "      <td>...</td>\n",
       "      <td>+</td>\n",
       "      <td>b0005</td>\n",
       "      <td>yaaX</td>\n",
       "      <td>predicted protein</td>\n",
       "      <td>Not in COGs</td>\n",
       "      <td>5234</td>\n",
       "      <td>5530</td>\n",
       "      <td>346</td>\n",
       "      <td>GAATATGGCTTGTTCCTCGGCACCGCGCATCCGGCGAAATTTAAAG...</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>3</td>\n",
       "      <td>0</td>\n",
       "      <td>3</td>\n",
       "      <td>yaaX</td>\n",
       "      <td>ATCTATTCATTATCTCAATCAGGCCGGGTTTGCTTTTATGCAGCCC...</td>\n",
       "      <td>5021</td>\n",
       "      <td>5233</td>\n",
       "      <td>213</td>\n",
       "      <td>thrC</td>\n",
       "      <td>Codirectional+</td>\n",
       "      <td>NC_000913</td>\n",
       "      <td>...</td>\n",
       "      <td>+</td>\n",
       "      <td>b0005</td>\n",
       "      <td>yaaX</td>\n",
       "      <td>predicted protein</td>\n",
       "      <td>Not in COGs</td>\n",
       "      <td>5234</td>\n",
       "      <td>5530</td>\n",
       "      <td>303</td>\n",
       "      <td>AAGAGAGCGTGGAAGCGATTCTCGGTGAAACGTTGGATCTGCCAAA...</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>4</td>\n",
       "      <td>2</td>\n",
       "      <td>6</td>\n",
       "      <td>yaaJ</td>\n",
       "      <td>TGATAGTATTTCTCTTTAAACAGCTTGTTAGGGGGATGTAACCGGT...</td>\n",
       "      <td>7960</td>\n",
       "      <td>8237</td>\n",
       "      <td>278</td>\n",
       "      <td>talB</td>\n",
       "      <td>Divergent</td>\n",
       "      <td>NC_000913</td>\n",
       "      <td>...</td>\n",
       "      <td>-</td>\n",
       "      <td>b0007</td>\n",
       "      <td>yaaJ</td>\n",
       "      <td>YaaJ alanine AGSS transporter</td>\n",
       "      <td>Amino acid transport and metabolism</td>\n",
       "      <td>6529</td>\n",
       "      <td>7959</td>\n",
       "      <td>209</td>\n",
       "      <td>CAGCTTGTTAGGGGGATGTAACCGGTCTGCCCTGATGATATCACGA...</td>\n",
       "      <td>TGATAGTATTTCTCTTTAAACAGCTTGTTAGGGGGATGTAACCGGT...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 22 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "   Unnamed: 0  IS Number  Gene  \\\n",
       "0           0          3  yaaX   \n",
       "1           0          3  yaaX   \n",
       "2           0          3  yaaX   \n",
       "3           0          3  yaaX   \n",
       "4           2          6  yaaJ   \n",
       "\n",
       "                                   Upstream Sequence  L_END  R_END  Length  \\\n",
       "0  ATCTATTCATTATCTCAATCAGGCCGGGTTTGCTTTTATGCAGCCC...   5021   5233     213   \n",
       "1  ATCTATTCATTATCTCAATCAGGCCGGGTTTGCTTTTATGCAGCCC...   5021   5233     213   \n",
       "2  ATCTATTCATTATCTCAATCAGGCCGGGTTTGCTTTTATGCAGCCC...   5021   5233     213   \n",
       "3  ATCTATTCATTATCTCAATCAGGCCGGGTTTGCTTTTATGCAGCCC...   5021   5233     213   \n",
       "4  TGATAGTATTTCTCTTTAAACAGCTTGTTAGGGGGATGTAACCGGT...   7960   8237     278   \n",
       "\n",
       "  Adjacent Gene     Orientation Chromosome  ...  Strand  Gene ID Gene name  \\\n",
       "0          thrC  Codirectional+  NC_000913  ...       +    b0005      yaaX   \n",
       "1          thrC  Codirectional+  NC_000913  ...       +    b0005      yaaX   \n",
       "2          thrC  Codirectional+  NC_000913  ...       +    b0005      yaaX   \n",
       "3          thrC  Codirectional+  NC_000913  ...       +    b0005      yaaX   \n",
       "4          talB       Divergent  NC_000913  ...       -    b0007      yaaJ   \n",
       "\n",
       "                    Gene product                         COG function  \\\n",
       "0              predicted protein                          Not in COGs   \n",
       "1              predicted protein                          Not in COGs   \n",
       "2              predicted protein                          Not in COGs   \n",
       "3              predicted protein                          Not in COGs   \n",
       "4  YaaJ alanine AGSS transporter  Amino acid transport and metabolism   \n",
       "\n",
       "  Gene start Gene end  5' UTR length  \\\n",
       "0       5234     5530            448   \n",
       "1       5234     5530            354   \n",
       "2       5234     5530            346   \n",
       "3       5234     5530            303   \n",
       "4       6529     7959            209   \n",
       "\n",
       "              Sequence -50 nt upstream + TSS (51 nt)  \\\n",
       "0  ACCACGCAACAGACAATGCGTGAGTTAAAAGAACTGGGCTACACTT...   \n",
       "1  ATCCAGGCGAATATGGCTTGTTCCTCGGCACCGCGCATCCGGCGAA...   \n",
       "2  GAATATGGCTTGTTCCTCGGCACCGCGCATCCGGCGAAATTTAAAG...   \n",
       "3  AAGAGAGCGTGGAAGCGATTCTCGGTGAAACGTTGGATCTGCCAAA...   \n",
       "4  CAGCTTGTTAGGGGGATGTAACCGGTCTGCCCTGATGATATCACGA...   \n",
       "\n",
       "                                Sequence without UTR  \n",
       "0                                                     \n",
       "1                                                     \n",
       "2                                                     \n",
       "3                                                     \n",
       "4  TGATAGTATTTCTCTTTAAACAGCTTGTTAGGGGGATGTAACCGGT...  \n",
       "\n",
       "[5 rows x 22 columns]"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "def cut_utr(ups_seq, utr_len, strand):\n",
    "    # if the utr_len is > ups_seq len, throw it out\n",
    "    if utr_len > len(ups_seq):\n",
    "        return ''\n",
    "    \n",
    "    # there are different off by one erros (presented by\n",
    "    # the data) depending on which strand we use\n",
    "    if strand == '+':\n",
    "        res = ups_seq[0:len(ups_seq) - utr_len - 1]\n",
    "    if strand == '-':\n",
    "        res = ups_seq[0:len(ups_seq) - utr_len + 1]\n",
    "    return res\n",
    "\n",
    "df_merged_cut = df_merged.copy()\n",
    "df_merged_cut['Sequence without UTR'] = df_merged.apply(\n",
    "    lambda row: cut_utr(row['Upstream Sequence'], row['5\\' UTR length'], row['Strand']), axis=1)\n",
    "\n",
    "df_merged_cut.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "**Note: Cases where UTR len > upstream sequence len.** For some genes at UTRs, the 5' UTR length is longer than the actual upstream sequence. These therefore have no sequence after we cut out the UTR and so must be dropped from the data set."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "1394\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Unnamed: 0</th>\n",
       "      <th>IS Number</th>\n",
       "      <th>Gene</th>\n",
       "      <th>Upstream Sequence</th>\n",
       "      <th>L_END</th>\n",
       "      <th>R_END</th>\n",
       "      <th>Length</th>\n",
       "      <th>Adjacent Gene</th>\n",
       "      <th>Orientation</th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>...</th>\n",
       "      <th>Strand</th>\n",
       "      <th>Gene ID</th>\n",
       "      <th>Gene name</th>\n",
       "      <th>Gene product</th>\n",
       "      <th>COG function</th>\n",
       "      <th>Gene start</th>\n",
       "      <th>Gene end</th>\n",
       "      <th>5' UTR length</th>\n",
       "      <th>Sequence -50 nt upstream + TSS (51 nt)</th>\n",
       "      <th>Sequence without UTR</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <td>4</td>\n",
       "      <td>2</td>\n",
       "      <td>6</td>\n",
       "      <td>yaaJ</td>\n",
       "      <td>TGATAGTATTTCTCTTTAAACAGCTTGTTAGGGGGATGTAACCGGT...</td>\n",
       "      <td>7960</td>\n",
       "      <td>8237</td>\n",
       "      <td>278</td>\n",
       "      <td>talB</td>\n",
       "      <td>Divergent</td>\n",
       "      <td>NC_000913</td>\n",
       "      <td>...</td>\n",
       "      <td>-</td>\n",
       "      <td>b0007</td>\n",
       "      <td>yaaJ</td>\n",
       "      <td>YaaJ alanine AGSS transporter</td>\n",
       "      <td>Amino acid transport and metabolism</td>\n",
       "      <td>6529</td>\n",
       "      <td>7959</td>\n",
       "      <td>209</td>\n",
       "      <td>CAGCTTGTTAGGGGGATGTAACCGGTCTGCCCTGATGATATCACGA...</td>\n",
       "      <td>TGATAGTATTTCTCTTTAAACAGCTTGTTAGGGGGATGTAACCGGT...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>6</td>\n",
       "      <td>3</td>\n",
       "      <td>6</td>\n",
       "      <td>talB</td>\n",
       "      <td>CATATCCCTCTTATTGCCGGTCGCGATGACTTTCCTGTGTAAACGT...</td>\n",
       "      <td>7960</td>\n",
       "      <td>8237</td>\n",
       "      <td>278</td>\n",
       "      <td>yaaJ</td>\n",
       "      <td>Divergent</td>\n",
       "      <td>NC_000913</td>\n",
       "      <td>...</td>\n",
       "      <td>+</td>\n",
       "      <td>b0008</td>\n",
       "      <td>talB</td>\n",
       "      <td>transaldolase B</td>\n",
       "      <td>Carbohydrate transport and metabolism</td>\n",
       "      <td>8238</td>\n",
       "      <td>9191</td>\n",
       "      <td>47</td>\n",
       "      <td>ACCGTCTTGTCGGCGGTTGCGCTGACGTTGCGTCGTGATATCATCA...</td>\n",
       "      <td>CATATCCCTCTTATTGCCGGTCGCGATGACTTTCCTGTGTAAACGT...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>8</td>\n",
       "      <td>6</td>\n",
       "      <td>11</td>\n",
       "      <td>dnaK</td>\n",
       "      <td>AGGATTCTCTTAGTGGGAAGAGGTAGGGGGATGAATACCCACTAGT...</td>\n",
       "      <td>11787</td>\n",
       "      <td>12162</td>\n",
       "      <td>376</td>\n",
       "      <td>yaaI</td>\n",
       "      <td>Divergent</td>\n",
       "      <td>NC_000913</td>\n",
       "      <td>...</td>\n",
       "      <td>+</td>\n",
       "      <td>b0014</td>\n",
       "      <td>dnaK</td>\n",
       "      <td>chaperone Hsp70, DNA biosynthesis, autoregulat...</td>\n",
       "      <td>Posttranslational modification, protein turnov...</td>\n",
       "      <td>12163</td>\n",
       "      <td>14079</td>\n",
       "      <td>6</td>\n",
       "      <td>CCCCTATTACAGACTCACAACCACATGATGACCGAATATATAGTGG...</td>\n",
       "      <td>AGGATTCTCTTAGTGGGAAGAGGTAGGGGGATGAATACCCACTAGT...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>11</td>\n",
       "      <td>10</td>\n",
       "      <td>17</td>\n",
       "      <td>nhaA</td>\n",
       "      <td>TATCTGCCGTTCAGCTAATGCCTGAGACAGACAGCCTCAAGCACCC...</td>\n",
       "      <td>17007</td>\n",
       "      <td>17488</td>\n",
       "      <td>482</td>\n",
       "      <td>sokC</td>\n",
       "      <td>Codirectional+</td>\n",
       "      <td>NC_000913</td>\n",
       "      <td>...</td>\n",
       "      <td>+</td>\n",
       "      <td>b0019</td>\n",
       "      <td>nhaA</td>\n",
       "      <td>NhaA sodium/proton transporter</td>\n",
       "      <td>Inorganic ion transport and metabolism</td>\n",
       "      <td>17489</td>\n",
       "      <td>18655</td>\n",
       "      <td>265</td>\n",
       "      <td>TCATCGCAATTATTGACGACAAGCTGGATTATTTTTGAAATATTGG...</td>\n",
       "      <td>TATCTGCCGTTCAGCTAATGCCTGAGACAGACAGCCTCAAGCACCC...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>12</td>\n",
       "      <td>12</td>\n",
       "      <td>22</td>\n",
       "      <td>rpsT</td>\n",
       "      <td>TTAACGGCGCTTATTTGCACAAATCCATTGACAAAAGAAGGCTAAA...</td>\n",
       "      <td>21079</td>\n",
       "      <td>21180</td>\n",
       "      <td>102</td>\n",
       "      <td>yaaY</td>\n",
       "      <td>Divergent</td>\n",
       "      <td>NC_000913</td>\n",
       "      <td>...</td>\n",
       "      <td>-</td>\n",
       "      <td>b0023</td>\n",
       "      <td>rpsT</td>\n",
       "      <td>30S ribosomal subunit protein S20</td>\n",
       "      <td>Translation</td>\n",
       "      <td>20815</td>\n",
       "      <td>21078</td>\n",
       "      <td>40</td>\n",
       "      <td>TTTGCACAAATCCATTGACAAAAGAAGGCTAAAAGGGCATATTCCT...</td>\n",
       "      <td>TTAACGGCGCTTATTTGCACAAATCCATTGACAAAAGAAGGCTAAA...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 22 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "    Unnamed: 0  IS Number  Gene  \\\n",
       "4            2          6  yaaJ   \n",
       "6            3          6  talB   \n",
       "8            6         11  dnaK   \n",
       "11          10         17  nhaA   \n",
       "12          12         22  rpsT   \n",
       "\n",
       "                                    Upstream Sequence  L_END  R_END  Length  \\\n",
       "4   TGATAGTATTTCTCTTTAAACAGCTTGTTAGGGGGATGTAACCGGT...   7960   8237     278   \n",
       "6   CATATCCCTCTTATTGCCGGTCGCGATGACTTTCCTGTGTAAACGT...   7960   8237     278   \n",
       "8   AGGATTCTCTTAGTGGGAAGAGGTAGGGGGATGAATACCCACTAGT...  11787  12162     376   \n",
       "11  TATCTGCCGTTCAGCTAATGCCTGAGACAGACAGCCTCAAGCACCC...  17007  17488     482   \n",
       "12  TTAACGGCGCTTATTTGCACAAATCCATTGACAAAAGAAGGCTAAA...  21079  21180     102   \n",
       "\n",
       "   Adjacent Gene     Orientation Chromosome  ...  Strand  Gene ID Gene name  \\\n",
       "4           talB       Divergent  NC_000913  ...       -    b0007      yaaJ   \n",
       "6           yaaJ       Divergent  NC_000913  ...       +    b0008      talB   \n",
       "8           yaaI       Divergent  NC_000913  ...       +    b0014      dnaK   \n",
       "11          sokC  Codirectional+  NC_000913  ...       +    b0019      nhaA   \n",
       "12          yaaY       Divergent  NC_000913  ...       -    b0023      rpsT   \n",
       "\n",
       "                                         Gene product  \\\n",
       "4                       YaaJ alanine AGSS transporter   \n",
       "6                                     transaldolase B   \n",
       "8   chaperone Hsp70, DNA biosynthesis, autoregulat...   \n",
       "11                     NhaA sodium/proton transporter   \n",
       "12                  30S ribosomal subunit protein S20   \n",
       "\n",
       "                                         COG function Gene start Gene end  \\\n",
       "4                 Amino acid transport and metabolism       6529     7959   \n",
       "6               Carbohydrate transport and metabolism       8238     9191   \n",
       "8   Posttranslational modification, protein turnov...      12163    14079   \n",
       "11             Inorganic ion transport and metabolism      17489    18655   \n",
       "12                                        Translation      20815    21078   \n",
       "\n",
       "    5' UTR length             Sequence -50 nt upstream + TSS (51 nt)  \\\n",
       "4             209  CAGCTTGTTAGGGGGATGTAACCGGTCTGCCCTGATGATATCACGA...   \n",
       "6              47  ACCGTCTTGTCGGCGGTTGCGCTGACGTTGCGTCGTGATATCATCA...   \n",
       "8               6  CCCCTATTACAGACTCACAACCACATGATGACCGAATATATAGTGG...   \n",
       "11            265  TCATCGCAATTATTGACGACAAGCTGGATTATTTTTGAAATATTGG...   \n",
       "12             40  TTTGCACAAATCCATTGACAAAAGAAGGCTAAAAGGGCATATTCCT...   \n",
       "\n",
       "                                 Sequence without UTR  \n",
       "4   TGATAGTATTTCTCTTTAAACAGCTTGTTAGGGGGATGTAACCGGT...  \n",
       "6   CATATCCCTCTTATTGCCGGTCGCGATGACTTTCCTGTGTAAACGT...  \n",
       "8   AGGATTCTCTTAGTGGGAAGAGGTAGGGGGATGAATACCCACTAGT...  \n",
       "11  TATCTGCCGTTCAGCTAATGCCTGAGACAGACAGCCTCAAGCACCC...  \n",
       "12  TTAACGGCGCTTATTTGCACAAATCCATTGACAAAAGAAGGCTAAA...  \n",
       "\n",
       "[5 rows x 22 columns]"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_merged_cut['Length of Sequence Without UTR'] = df_merged_cut['Sequence without UTR'].apply(len)\n",
    "mask1 = df_merged_cut['Length of Sequence Without UTR'] > 0\n",
    "\n",
    "df_merged_cut = df_merged_cut[mask1]\n",
    "df_merged_cut = df_merged_cut.drop(columns=\"Length of Sequence Without UTR\")\n",
    "\n",
    "\n",
    "print(len(df_merged_cut))\n",
    "df_merged_cut.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<a id=\"div\"></a>\n",
    "### Divergent Case"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The divergent case is a lot trickier. We need to cut off the UTRs from both sides (\\*\\* NOTE: do not forget about the off by one error) for each TSS. This will create a lot more entries in our table. We also need to add another column which denotes the side of the TSS site of the other side that got cut."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "First we just grab the entries with divergent orientation"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Unnamed: 0</th>\n",
       "      <th>IS Number</th>\n",
       "      <th>Gene</th>\n",
       "      <th>Upstream Sequence</th>\n",
       "      <th>L_END</th>\n",
       "      <th>R_END</th>\n",
       "      <th>Length</th>\n",
       "      <th>Adjacent Gene</th>\n",
       "      <th>Orientation</th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>...</th>\n",
       "      <th>Strand</th>\n",
       "      <th>Gene ID</th>\n",
       "      <th>Gene name</th>\n",
       "      <th>Gene product</th>\n",
       "      <th>COG function</th>\n",
       "      <th>Gene start</th>\n",
       "      <th>Gene end</th>\n",
       "      <th>5' UTR length</th>\n",
       "      <th>Sequence -50 nt upstream + TSS (51 nt)</th>\n",
       "      <th>Sequence without UTR</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <td>4</td>\n",
       "      <td>2</td>\n",
       "      <td>6</td>\n",
       "      <td>yaaJ</td>\n",
       "      <td>TGATAGTATTTCTCTTTAAACAGCTTGTTAGGGGGATGTAACCGGT...</td>\n",
       "      <td>7960</td>\n",
       "      <td>8237</td>\n",
       "      <td>278</td>\n",
       "      <td>talB</td>\n",
       "      <td>Divergent</td>\n",
       "      <td>NC_000913</td>\n",
       "      <td>...</td>\n",
       "      <td>-</td>\n",
       "      <td>b0007</td>\n",
       "      <td>yaaJ</td>\n",
       "      <td>YaaJ alanine AGSS transporter</td>\n",
       "      <td>Amino acid transport and metabolism</td>\n",
       "      <td>6529</td>\n",
       "      <td>7959</td>\n",
       "      <td>209</td>\n",
       "      <td>CAGCTTGTTAGGGGGATGTAACCGGTCTGCCCTGATGATATCACGA...</td>\n",
       "      <td>TGATAGTATTTCTCTTTAAACAGCTTGTTAGGGGGATGTAACCGGT...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>6</td>\n",
       "      <td>3</td>\n",
       "      <td>6</td>\n",
       "      <td>talB</td>\n",
       "      <td>CATATCCCTCTTATTGCCGGTCGCGATGACTTTCCTGTGTAAACGT...</td>\n",
       "      <td>7960</td>\n",
       "      <td>8237</td>\n",
       "      <td>278</td>\n",
       "      <td>yaaJ</td>\n",
       "      <td>Divergent</td>\n",
       "      <td>NC_000913</td>\n",
       "      <td>...</td>\n",
       "      <td>+</td>\n",
       "      <td>b0008</td>\n",
       "      <td>talB</td>\n",
       "      <td>transaldolase B</td>\n",
       "      <td>Carbohydrate transport and metabolism</td>\n",
       "      <td>8238</td>\n",
       "      <td>9191</td>\n",
       "      <td>47</td>\n",
       "      <td>ACCGTCTTGTCGGCGGTTGCGCTGACGTTGCGTCGTGATATCATCA...</td>\n",
       "      <td>CATATCCCTCTTATTGCCGGTCGCGATGACTTTCCTGTGTAAACGT...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>8</td>\n",
       "      <td>6</td>\n",
       "      <td>11</td>\n",
       "      <td>dnaK</td>\n",
       "      <td>AGGATTCTCTTAGTGGGAAGAGGTAGGGGGATGAATACCCACTAGT...</td>\n",
       "      <td>11787</td>\n",
       "      <td>12162</td>\n",
       "      <td>376</td>\n",
       "      <td>yaaI</td>\n",
       "      <td>Divergent</td>\n",
       "      <td>NC_000913</td>\n",
       "      <td>...</td>\n",
       "      <td>+</td>\n",
       "      <td>b0014</td>\n",
       "      <td>dnaK</td>\n",
       "      <td>chaperone Hsp70, DNA biosynthesis, autoregulat...</td>\n",
       "      <td>Posttranslational modification, protein turnov...</td>\n",
       "      <td>12163</td>\n",
       "      <td>14079</td>\n",
       "      <td>6</td>\n",
       "      <td>CCCCTATTACAGACTCACAACCACATGATGACCGAATATATAGTGG...</td>\n",
       "      <td>AGGATTCTCTTAGTGGGAAGAGGTAGGGGGATGAATACCCACTAGT...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>12</td>\n",
       "      <td>12</td>\n",
       "      <td>22</td>\n",
       "      <td>rpsT</td>\n",
       "      <td>TTAACGGCGCTTATTTGCACAAATCCATTGACAAAAGAAGGCTAAA...</td>\n",
       "      <td>21079</td>\n",
       "      <td>21180</td>\n",
       "      <td>102</td>\n",
       "      <td>yaaY</td>\n",
       "      <td>Divergent</td>\n",
       "      <td>NC_000913</td>\n",
       "      <td>...</td>\n",
       "      <td>-</td>\n",
       "      <td>b0023</td>\n",
       "      <td>rpsT</td>\n",
       "      <td>30S ribosomal subunit protein S20</td>\n",
       "      <td>Translation</td>\n",
       "      <td>20815</td>\n",
       "      <td>21078</td>\n",
       "      <td>40</td>\n",
       "      <td>TTTGCACAAATCCATTGACAAAAGAAGGCTAAAAGGGCATATTCCT...</td>\n",
       "      <td>TTAACGGCGCTTATTTGCACAAATCCATTGACAAAAGAAGGCTAAA...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>20</td>\n",
       "      <td>24</td>\n",
       "      <td>53</td>\n",
       "      <td>lptD</td>\n",
       "      <td>TATATTCCCCAAATCGACACACGGATATCAGGGCTATCTCCCACAA...</td>\n",
       "      <td>57110</td>\n",
       "      <td>57363</td>\n",
       "      <td>254</td>\n",
       "      <td>djlA</td>\n",
       "      <td>Divergent</td>\n",
       "      <td>NC_000913</td>\n",
       "      <td>...</td>\n",
       "      <td>-</td>\n",
       "      <td>b0054</td>\n",
       "      <td>lptD</td>\n",
       "      <td>outer membrane LPS assembly complex</td>\n",
       "      <td>Cell wall/membrane biogenesis</td>\n",
       "      <td>54755</td>\n",
       "      <td>57109</td>\n",
       "      <td>47</td>\n",
       "      <td>GGACAGGATTAACACTAGCGTAGAGATGACGAGTACGTTAGTCTCT...</td>\n",
       "      <td>TATATTCCCCAAATCGACACACGGATATCAGGGCTATCTCCCACAA...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 22 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "    Unnamed: 0  IS Number  Gene  \\\n",
       "4            2          6  yaaJ   \n",
       "6            3          6  talB   \n",
       "8            6         11  dnaK   \n",
       "12          12         22  rpsT   \n",
       "20          24         53  lptD   \n",
       "\n",
       "                                    Upstream Sequence  L_END  R_END  Length  \\\n",
       "4   TGATAGTATTTCTCTTTAAACAGCTTGTTAGGGGGATGTAACCGGT...   7960   8237     278   \n",
       "6   CATATCCCTCTTATTGCCGGTCGCGATGACTTTCCTGTGTAAACGT...   7960   8237     278   \n",
       "8   AGGATTCTCTTAGTGGGAAGAGGTAGGGGGATGAATACCCACTAGT...  11787  12162     376   \n",
       "12  TTAACGGCGCTTATTTGCACAAATCCATTGACAAAAGAAGGCTAAA...  21079  21180     102   \n",
       "20  TATATTCCCCAAATCGACACACGGATATCAGGGCTATCTCCCACAA...  57110  57363     254   \n",
       "\n",
       "   Adjacent Gene Orientation Chromosome  ...  Strand  Gene ID Gene name  \\\n",
       "4           talB   Divergent  NC_000913  ...       -    b0007      yaaJ   \n",
       "6           yaaJ   Divergent  NC_000913  ...       +    b0008      talB   \n",
       "8           yaaI   Divergent  NC_000913  ...       +    b0014      dnaK   \n",
       "12          yaaY   Divergent  NC_000913  ...       -    b0023      rpsT   \n",
       "20          djlA   Divergent  NC_000913  ...       -    b0054      lptD   \n",
       "\n",
       "                                         Gene product  \\\n",
       "4                       YaaJ alanine AGSS transporter   \n",
       "6                                     transaldolase B   \n",
       "8   chaperone Hsp70, DNA biosynthesis, autoregulat...   \n",
       "12                  30S ribosomal subunit protein S20   \n",
       "20                outer membrane LPS assembly complex   \n",
       "\n",
       "                                         COG function Gene start Gene end  \\\n",
       "4                 Amino acid transport and metabolism       6529     7959   \n",
       "6               Carbohydrate transport and metabolism       8238     9191   \n",
       "8   Posttranslational modification, protein turnov...      12163    14079   \n",
       "12                                        Translation      20815    21078   \n",
       "20                      Cell wall/membrane biogenesis      54755    57109   \n",
       "\n",
       "    5' UTR length             Sequence -50 nt upstream + TSS (51 nt)  \\\n",
       "4             209  CAGCTTGTTAGGGGGATGTAACCGGTCTGCCCTGATGATATCACGA...   \n",
       "6              47  ACCGTCTTGTCGGCGGTTGCGCTGACGTTGCGTCGTGATATCATCA...   \n",
       "8               6  CCCCTATTACAGACTCACAACCACATGATGACCGAATATATAGTGG...   \n",
       "12             40  TTTGCACAAATCCATTGACAAAAGAAGGCTAAAAGGGCATATTCCT...   \n",
       "20             47  GGACAGGATTAACACTAGCGTAGAGATGACGAGTACGTTAGTCTCT...   \n",
       "\n",
       "                                 Sequence without UTR  \n",
       "4   TGATAGTATTTCTCTTTAAACAGCTTGTTAGGGGGATGTAACCGGT...  \n",
       "6   CATATCCCTCTTATTGCCGGTCGCGATGACTTTCCTGTGTAAACGT...  \n",
       "8   AGGATTCTCTTAGTGGGAAGAGGTAGGGGGATGAATACCCACTAGT...  \n",
       "12  TTAACGGCGCTTATTTGCACAAATCCATTGACAAAAGAAGGCTAAA...  \n",
       "20  TATATTCCCCAAATCGACACACGGATATCAGGGCTATCTCCCACAA...  \n",
       "\n",
       "[5 rows x 22 columns]"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_div = df_merged_cut[df_merged_cut['Orientation'] == 'Divergent']\n",
    "# df_div = df_div[(df_div['Gene'] == 'yaaJ') | (df_div['Gene'] == 'talB')] # uncomment to see simple case\n",
    "df_div.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Unnamed: 0</th>\n",
       "      <th>IS Number</th>\n",
       "      <th>Gene</th>\n",
       "      <th>Upstream Sequence</th>\n",
       "      <th>L_END</th>\n",
       "      <th>R_END</th>\n",
       "      <th>Length</th>\n",
       "      <th>Adjacent Gene</th>\n",
       "      <th>Orientation</th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>...</th>\n",
       "      <th>Strand</th>\n",
       "      <th>Gene ID</th>\n",
       "      <th>Gene name</th>\n",
       "      <th>Gene product</th>\n",
       "      <th>COG function</th>\n",
       "      <th>Gene start</th>\n",
       "      <th>Gene end</th>\n",
       "      <th>5' UTR length</th>\n",
       "      <th>Sequence -50 nt upstream + TSS (51 nt)</th>\n",
       "      <th>Sequence without UTR</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <td>4</td>\n",
       "      <td>2</td>\n",
       "      <td>6</td>\n",
       "      <td>yaaJ</td>\n",
       "      <td>TGATAGTATTTCTCTTTAAACAGCTTGTTAGGGGGATGTAACCGGT...</td>\n",
       "      <td>7960</td>\n",
       "      <td>8237</td>\n",
       "      <td>278</td>\n",
       "      <td>talB</td>\n",
       "      <td>Divergent</td>\n",
       "      <td>NC_000913</td>\n",
       "      <td>...</td>\n",
       "      <td>-</td>\n",
       "      <td>b0007</td>\n",
       "      <td>yaaJ</td>\n",
       "      <td>YaaJ alanine AGSS transporter</td>\n",
       "      <td>Amino acid transport and metabolism</td>\n",
       "      <td>6529</td>\n",
       "      <td>7959</td>\n",
       "      <td>209</td>\n",
       "      <td>CAGCTTGTTAGGGGGATGTAACCGGTCTGCCCTGATGATATCACGA...</td>\n",
       "      <td>TGATAGTATTTCTCTTTAAACAGCTTGTTAGGGGGATGTAACCGGT...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>6</td>\n",
       "      <td>3</td>\n",
       "      <td>6</td>\n",
       "      <td>talB</td>\n",
       "      <td>CATATCCCTCTTATTGCCGGTCGCGATGACTTTCCTGTGTAAACGT...</td>\n",
       "      <td>7960</td>\n",
       "      <td>8237</td>\n",
       "      <td>278</td>\n",
       "      <td>yaaJ</td>\n",
       "      <td>Divergent</td>\n",
       "      <td>NC_000913</td>\n",
       "      <td>...</td>\n",
       "      <td>+</td>\n",
       "      <td>b0008</td>\n",
       "      <td>talB</td>\n",
       "      <td>transaldolase B</td>\n",
       "      <td>Carbohydrate transport and metabolism</td>\n",
       "      <td>8238</td>\n",
       "      <td>9191</td>\n",
       "      <td>47</td>\n",
       "      <td>ACCGTCTTGTCGGCGGTTGCGCTGACGTTGCGTCGTGATATCATCA...</td>\n",
       "      <td>CATATCCCTCTTATTGCCGGTCGCGATGACTTTCCTGTGTAAACGT...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>2 rows × 22 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "   Unnamed: 0  IS Number  Gene  \\\n",
       "4           2          6  yaaJ   \n",
       "6           3          6  talB   \n",
       "\n",
       "                                   Upstream Sequence  L_END  R_END  Length  \\\n",
       "4  TGATAGTATTTCTCTTTAAACAGCTTGTTAGGGGGATGTAACCGGT...   7960   8237     278   \n",
       "6  CATATCCCTCTTATTGCCGGTCGCGATGACTTTCCTGTGTAAACGT...   7960   8237     278   \n",
       "\n",
       "  Adjacent Gene Orientation Chromosome  ...  Strand  Gene ID Gene name  \\\n",
       "4          talB   Divergent  NC_000913  ...       -    b0007      yaaJ   \n",
       "6          yaaJ   Divergent  NC_000913  ...       +    b0008      talB   \n",
       "\n",
       "                    Gene product                           COG function  \\\n",
       "4  YaaJ alanine AGSS transporter    Amino acid transport and metabolism   \n",
       "6                transaldolase B  Carbohydrate transport and metabolism   \n",
       "\n",
       "  Gene start Gene end  5' UTR length  \\\n",
       "4       6529     7959            209   \n",
       "6       8238     9191             47   \n",
       "\n",
       "              Sequence -50 nt upstream + TSS (51 nt)  \\\n",
       "4  CAGCTTGTTAGGGGGATGTAACCGGTCTGCCCTGATGATATCACGA...   \n",
       "6  ACCGTCTTGTCGGCGGTTGCGCTGACGTTGCGTCGTGATATCATCA...   \n",
       "\n",
       "                                Sequence without UTR  \n",
       "4  TGATAGTATTTCTCTTTAAACAGCTTGTTAGGGGGATGTAACCGGT...  \n",
       "6  CATATCCCTCTTATTGCCGGTCGCGATGACTTTCCTGTGTAAACGT...  \n",
       "\n",
       "[2 rows x 22 columns]"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_div[(df_div['Gene'] == 'talB') | (df_div['Gene'] == 'yaaJ')]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Next we cut the front ends off of the sequences, and store the adjacent genes TSS that was used to cut off the front of the sequence"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "# function for cutting the front of the utrs for the divergent case\n",
    "def cut_utr_front(ups_seq, utr_len, strand):\n",
    "    # throw it out if it would just cut the whole thing up\n",
    "    if utr_len > len(ups_seq):\n",
    "        return ''\n",
    "    \n",
    "    # off by one error, change the end based on\n",
    "    # strand direction of the other strand\n",
    "    if strand == \"+\":\n",
    "        res = ups_seq[utr_len+1:]\n",
    "    else:\n",
    "        res = ups_seq[utr_len-1:]\n",
    "    return res"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "This part takes a while, hence the Run variable. Set it to True if you need to run the code, otherwise it will load in already existing data."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "Run = True\n",
    "pd.options.mode.chained_assignment = None  # default='warn'\n",
    "\n",
    "num_missing_adj_entries = 0\n",
    "\n",
    "if Run:\n",
    "    div_lst = df_div.to_dict('records')  # a list of dictionaries of the divergent entries, easier to iterate\n",
    "\n",
    "    div_exp = pd.DataFrame()  # data structure to store expanded entries\n",
    "\n",
    "    for r in div_lst:\n",
    "        # get neccesary data\n",
    "        gene = r['Gene']\n",
    "        utr_len = r['5\\' UTR length']\n",
    "        adj_gene = r['Adjacent Gene']\n",
    "        TSS = r['TSS position']  \n",
    "        strand = r['Strand']\n",
    "\n",
    "\n",
    "        # get the rows that have the adjacent gene\n",
    "        adj_genes = df_div[df_div['Gene'] == adj_gene]\n",
    "        if len(adj_genes) == 0:\n",
    "            num_missing_adj_entries += 1\n",
    "        adj_genes['Sequence without UTR'] = df_merged_cut.apply(\n",
    "            lambda row: cut_utr_front(row['Sequence without UTR'], utr_len, strand), axis=1)                                                    \n",
    "        adj_genes['adj_TSS'] = TSS\n",
    "        div_exp = pd.concat([div_exp, adj_genes])       \n",
    "    div_exp = div_exp.dropna()\n",
    "    div_exp.to_excel(\"./export/div_exp_int.xlsx\")\n",
    "else:\n",
    "    div_exp =  pd.read_excel(\"./export/div_exp_int.xlsx\")\n",
    "div_exp.head()    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "# 1 - num_missing_adj_entries/len(div_lst), 57, \"percent of entries lost because adjacent gene was missing\""
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Drop all entries where there no longer is a sequence"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "def mod_len(string):\n",
    "    if type(string) == float:\n",
    "        return 0\n",
    "    else:\n",
    "        return len(string)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "# remove entries without the length\n",
    "print(len(div_exp))\n",
    "div_exp['Length of Sequence Without UTR'] = div_exp['Sequence without UTR'].apply(mod_len)\n",
    "mask1 = div_exp['Length of Sequence Without UTR'] > 0\n",
    "\n",
    "div_exp = div_exp[mask1]\n",
    "div_exp = div_exp.drop(columns=\"Length of Sequence Without UTR\")\n",
    "print(len(div_exp))\n",
    "# div_exp"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "1 - 457/525, \"divergent genes lost because both ends were cut off\""
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Concat this dataframe with all of our codirectional data entries, and sort by IS Number so they're in the same order as before. (I don't know if this is the final ordering we want everything in, but should be easy to change)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "df_co = df_merged_cut[df_merged_cut['Orientation'] != 'Divergent']\n",
    "df_co[\"adj_TSS\"] = np.nan\n",
    "df_merged_cut_div = pd.concat([df_co, div_exp])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "df_merged_cut_div = df_merged_cut_div.sort_values(by=['IS Number'])\n",
    "df_merged_cut_div.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "len(df_merged_cut_div), len(df_co), len(div_exp)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "**Note: Data lost.** It seems likely that some sequences got cut off on both ends so that there was overlap and so no sequence covered by both genes. In addition, the adjacent gene for some genes was missing from the dataset (likely lost in a previous step or merge). An example of this is seen with the gene dnaK. It's adjacent gene is yaaI, which is not present in the data set."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<a id=\"reorder\"></a>\n",
    "### Reordering"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Start by grabbing the columns we want"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "df_reordered = df_merged_cut_div[['IS Number', \n",
    "             'Gene', \n",
    "             'Gene ID', \n",
    "             'Gene product', \n",
    "             'COG function', \n",
    "             'L_END', \n",
    "             'R_END', \n",
    "             'Length', \n",
    "             'Upstream Sequence', \n",
    "             'Orientation', \n",
    "             'Strand', \n",
    "             'TSS position', \n",
    "             'TSS reads', \n",
    "             'Sequence without UTR', \n",
    "             'Sequence -50 nt upstream + TSS (51 nt)', \n",
    "             '5\\' UTR length',\n",
    "             'Gene start',\n",
    "             'Gene end',\n",
    "             'Adjacent Gene',\n",
    "             'adj_TSS'\n",
    "            ]]\n",
    "df_reordered.iloc[0:2]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Renaming the columns"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "df_reordered = df_reordered.rename(index=str, \n",
    "                   columns={\n",
    "                       \"L_END\": \"IS Left End\",\n",
    "                       \"R_END\": \"IS Right End\",\n",
    "                       \"Length\": \"IS Length\",\n",
    "                       \"Upstream Sequence\": \"Intergenic Sequence\",\n",
    "                       \"Sequence without UTR\": \"Intergenic Sequence no UTR\",\n",
    "                       \"adj_TSS\": \"Ajacent Gene TSS position\"                      \n",
    "                   })\n",
    "\n",
    "df_reordered.iloc[0:2]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<a id=\"export\"></a>\n",
    "### Exporting"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "df_reordered.to_excel(\"./20190717-IS-noUTRs-df1.xlsx\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "---\n",
    "<a id=\"test\"></a>\n",
    "## Tests\n",
    "<a id=\"co_test\"></a>\n",
    "### Codirectional Case\n",
    "For this case, we are just checking if the UTR got cut out correctly, checking it in benchling with ecocyc.\n",
    "\n",
    "Things to test:\n",
    "* Genes with a utr longer than the intergenic sequence were removed\n",
    "* Actually cut correctly"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "df_reordered.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "# all instances of yaaX have a longer 5' UTR than the intergenic sequence and so should have been removed\n",
    "assert(len(df_reordered[df_reordered['Gene'] == 'yaaX']) == 0)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<a id=\"plus\"></a>\n",
    "#### Codirectional +: nhaA"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Checking for a gene that did actually work (nhaA). Note that it should since the length of the intergenic sequence is 482 and the 5' UTR length is 265.0"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "co = df_reordered[df_reordered['Orientation'] != 'Divergent']\n",
    "co.head()\n",
    "nhaA = df_reordered[df_reordered['Gene'] == 'nhaA']\n",
    "nhaA"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "ups_seq = nhaA['Intergenic Sequence'][0]\n",
    "utr_len = nhaA['5\\' UTR length'][0]\n",
    "no_utr = nhaA['Intergenic Sequence no UTR'][0]\n",
    "ups_seq, utr_len, no_utr"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "I then checked this in benchling.\n",
    "* First I went to [ecocyc](https://ecocyc.org/gene?orgid=ECOLI&id=EG10652) to get the information for nhaA and pasted it into benchling\n",
    "* Assuming that the 5' UTR length was correct, I annotated it and found where the Intergenic Sequence no UTR should end\n",
    "* I then checked that the Intergenic Sequence no UTR generated by code matched\n",
    "\n",
    "It did! Success for + codirectional case\n",
    "\n",
    "Here is the benchling file I used for this test:\n",
    "\n",
    "https://benchling.com/s/seq-3Cy6xOezeaXP4FwYMMGQ"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Although this appears to be correct according to our information, the benchling data implies that this is the incorrect UTR length. On benchling it says the TSS should be at position 17317 or 17458 instead of 17224"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "17224 - 17317 "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<a id=\"minus\"></a>\n",
    "#### Codirectional -: polB\n",
    "After starting to test the divergent case, I realized there are different problems that arise from the codirectional - case. These come from different off by 1 errros"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "polB = co[co['Gene'] == 'polB']\n",
    "polB"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "ups_seq = polB['Intergenic Sequence'][0]\n",
    "utr_len = polB['5\\' UTR length'][0]\n",
    "no_utr = polB['Intergenic Sequence no UTR'][0]\n",
    "ups_seq, utr_len, no_utr"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "I then checked this in benchling.\n",
    "* I got the information for [polB](https://ecocyc.org/gene?orgid=ECOLI&id=EG10652) and put it into benchling\n",
    "* Assuming that the 5' UTR length was correct, I annotated it and found where the Intergenic Sequence no UTR should end\n",
    "* I then checked that the Intergenic Sequence no UTR generated by code matched\n",
    "\n",
    "Success for Codirectional -\n",
    "\n",
    "\n",
    "Here is the benchling file I used for this test:\n",
    "\n",
    "https://benchling.com/s/seq-IROLxENViRMKzNXJiLCI"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<a id=\"div_test\"></a>\n",
    "### Divergent Case: talB and yaaJ"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "We'll look at talB and yaaJ (seeing if they were both done correctly). Since they each only have one valid TSS each, we will examine this case since it is simple. I have checked earlier to make sure the combinatorics were working (ie the case with multiple TSSs), hence why I will check this simple case"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "div = df_reordered[df_reordered['Orientation'] == 'Divergent']\n",
    "div.head()\n",
    "talB = df_reordered[df_reordered['Gene'] == 'talB']\n",
    "yaaJ = df_reordered[df_reordered['Gene'] == 'yaaJ']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "talB_ups_seq = talB['Intergenic Sequence'][0]\n",
    "talB_utr_len = talB['5\\' UTR length'][0]\n",
    "talB_no_utr = talB['Intergenic Sequence no UTR'][0]\n",
    "\n",
    "yaaJ_ups_seq = yaaJ['Intergenic Sequence'][0]\n",
    "yaaJ_utr_len = yaaJ['5\\' UTR length'][0]\n",
    "yaaJ_no_utr = yaaJ['Intergenic Sequence no UTR'][0]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "the Intergenic Sequence and Intergenic Sequence no UTR of both genes should have the same length and be reverse compliments of eachother. This should catch errors in the data and off by one errors"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "assert(len(talB_ups_seq) == len(yaaJ_ups_seq)), \"talB and yaaJ should have the same length of intergenic sequence\"\n",
    "assert(talB_ups_seq == rev_comp(yaaJ_ups_seq)), \"talB's intergenic sequence should be the reverse compliment of yaaJ's intergenic sequence\"\n",
    "assert(len(yaaJ_no_utr) == len(talB_no_utr)), \"talB and yaaJ should have the same length of Intergenic Sequence no UTR\"\n",
    "assert(yaaJ_no_utr == rev_comp(talB_no_utr)), \"talB's intergenic sequence no UTR should be the reverse compliment of yaaJ's intergenic sequence no UTR\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "talB_no_utr, talB_utr_len, yaaJ_no_utr, yaaJ_utr_len, "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Finally we'll do a benchling check\n",
    "* Got the ecocyc info for [talB](https://ecocyc.org/gene?orgid=ECOLI&id=EG11556) and [yaaJ](https://ecocyc.org/gene?orgid=ECOLI&id=EG11555)\n",
    "* Assuming that the 5' UTR length was correct, I annotated them both.\n",
    "* I then checked that the Intergenic Sequence no UTR generated by code matched\n",
    "\n",
    "Here are the benchling files I used for this test:\n",
    "\n",
    "talB: https://benchling.com/s/seq-84PLtGcoFBE43orWdOIf\n",
    "\n",
    "yaaJ: https://benchling.com/s/seq-SdK8PctSCKrI96bWQMBo"
   ]
  }
 ],
 "metadata": {
  "anaconda-cloud": {},
  "kernelspec": {
   "display_name": "Python [conda root]",
   "language": "python",
   "name": "conda-root-py"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 2
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython2",
   "version": "2.7.13"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
